airflow/jobs/scheduler_job_runner.py

```diff
@@ -1844,6 +1895,58 @@ def _executor_to_tis(self, tis: list[TaskInstance]) -> dict[BaseExecutor, list[T
                 executor = str(ti.executor)
             else:
                 executor = None
-            _executor_to_tis[ExecutorLoader.load_executor(executor)].append(ti)
+
+            if executor_obj := self._try_to_load_executor(executor):
+                _executor_to_tis[executor_obj].append(ti)
+            else:
+                continue
 
         return _executor_to_tis
+
+    def _slots_free_for_tis(self, tis: list[TaskInstance]) -> set[TaskInstance]:
+        """Return TIs that we have slots available for."""
+        # First get a mapping of executor names to slots they have available
+        executor_to_slots_available: dict[ExecutorName, int] = {}
+        for executor in self.job.executors:
+            # All executors should have a name if they are initted from the executor_loader. But we need to
+            # check for None to make mypy happy.
+            if executor.name:
+                executor_to_slots_available[executor.name] = executor.slots_available
+            else:
+                raise AirflowException(f"Executor {executor} did not have a `name` field configured!")
+
+        # Loop through all the TIs we're scheduling for and add them to a set to be moved to queued if the
+        # executor they're assigned to has slots available.
+        tis_we_have_room_for = set()
+        for ti in tis:
+            if ti.executor:
+                executor = str(ti.executor)
+            else:
+                executor = None
```

Comment:
is this another case for if TYPE_CHECKING?

Reply:
some of this seems to be a duplicate of logic in `_executor_to_tis` wondering if there might be a way to reuse it?

Reply:
This is more type checking, but in this case the value really can be None, so the trick above of doing an assert won't work. In this case the mypy can't detect which of the types the value is because it arrives as a SA Column type. So here I'm casting it from that Column type to either str or None. Then _try_to_load_executor can handle either of those.

With that context would you like a change here?

Reply:
it's just a suggestion, take it or leave it :)

Reply:
Sounds good!